### PR TITLE
fix(step-functions): compare undefined and null

### LINKED ISSUE #441: Updating the logging configuration for ExpressStepFunction does not trigger a redeploy
Steps to reproduce:

1. Deploy an `ExpressStepFunction` with no logging configuration
2. Add a logging configuration.
3. Re-deploy

Expected outcome:

The Step Function should be updated with the new logging config.

Actual outcome:

The logging configuration remains unmodified.

### PR DESCRIPTION
**Pull request by thantos**
Fixes #441

`x === undefined` and other comparison patterns.

Root Cause:
* Bad typo (didn't `return`)
* Missing `undefined` case when comparing two variables or variable and a constant.
* Empty `or` logic returned true (should be false)

### CODE REVIEW

File: src/asl.ts [hunk 1]
```diff
@@ -5431,6 +5445,24 @@ export namespace ASLGraph {
...
+      if (ASLGraph.isJsonPath(right)) {
+        return ASL.and(
+          ASL.isNotPresent(left.jsonPath),
+          ASL.isNotPresent(right.jsonPath)
+        );
+      } else if (ASLGraph.isLiteralValue(right) && right.value === undefined) {
+        return ASL.isNotPresent(left.jsonPath);
+      }
+    }
+    return undefined;
```
Comment (sam-goodwin) [inline src/asl.ts:5463]: Is this case expected?

File: src/asl.ts [hunk 2]
```diff
@@ -5261,36 +5270,41 @@ export namespace ASLGraph {
...
     ) {
+      // left is a condition and right is any value
       if (ASLGraph.isConditionOutput(left)) {
         return (
           ASLGraph.booleanCompare(left, right, operator) ?? ASL.falseCondition()
         );
       }
+
+      // returns comparisons if there are valid comparisons to return.
+      // comparisons return undefined when the right is a literal and not the right
```
Comment (sam-goodwin) [inline src/asl.ts:5281]: > when the right is a literal and not the right

Typo?

File: src/asl.ts [hunk 3]
```diff
@@ -5261,36 +5270,41 @@ export namespace ASLGraph {
...
-            condition
-          )
-        );
-      }
-      return ASL.and(ASL.isPresent(left.jsonPath), condition);
+      return ASL.or(
+        // a === b while a and b are both not defined
+        ASLGraph.undefinedCompare(left, right, operator),
+        // a !== undefined && b !== undefined
+        ASL.and(
```
Comment (sam-goodwin) [inline src/asl.ts:5295]: I don't understand what the comments mean

**Review by sam-goodwin — APPROVED** [review]

File: src/asl.ts (hunk 1 above)
Comment (thantos) [reply src/asl.ts:5463]: Yeah, returns undefined when there is no comparison to add. If there are no comparisons at all, `false` is returned at runtime.

* `x === undefined` => returns conditional (`isNotPresent(x)`)
* `x === "a"` => return undefined
* `x === y` => return conditional (`isNotPresent(x) && isNotPresent(y)`) 
* `x < undefined` => return undefined

File: src/asl.ts (hunk 3 above)
Comment (thantos) [reply src/asl.ts:5295]: Woops, not relevant anymore.

File: src/asl.ts [hunk 4]
```diff
@@ -5261,42 +5270,59 @@ export namespace ASLGraph {
...
-            condition
-          )
-        );
-      }
-      return ASL.and(ASL.isPresent(left.jsonPath), condition);
+      return ASL.or(
+        // for ==/===
+        // if right is undefined literal, check isNotPresent(left)
+        // if right is a variable, check isNotPresent(left) && isNotPresent(right)
+        ASLGraph.undefinedCompare(left, right, operator),
```
Comment (sam-goodwin) [inline src/asl.ts:5296]: What happens if this returns `undefined`?

Comment (thantos) [reply src/asl.ts:5296]: We are trying to collect the comparisons we need, when the values are literals, we only do the comparisons of the type of the literal value. When the values are variable, we must do all of the comparisons to try to find a `true`.

So the `*compare` methods return a Condition when we need a runtime condition and `undefined` when we can assume the comparison is false.

Comment (sam-goodwin) [reply src/asl.ts]: I still don't understand what happens when `ASL.or(undefined, ..)`

Comment (thantos) [reply src/asl.ts]: ohh, `or` and `and` filter out undefined. if all undefined are passed to `or`, it returns falsy condition, if all undefined are passed to AND it returns a truthy condition.

**Pull request merged**